broadway_api/callbacks/worker.py

```diff
@@ -15,16 +18,88 @@ def worker_heartbeat_callback(settings):
     """
     heartbeat_timestamp = get_time()
     heartbeat_interval = settings["CONFIG"]["HEARTBEAT_INTERVAL"]
+    conn_map = settings["WS_CONN_MAP"]
 
     dao = WorkerNodeDao(settings)
+
     for node in dao.find_by_liveness(alive=True):
         if (
             heartbeat_timestamp - node.last_seen
         ).total_seconds() >= 2 * heartbeat_interval:
+            if node.use_ws and node.id in conn_map:
+                conn_map[node.id].close()
+
             _handle_lost_worker_node(settings, node)
 
+        elif node.use_ws and node.id in conn_map:
+            # send ping messages to all connections
+            # excepting pong response before the next checking point
+            conn_map[node.id].ping()
+
+
+def worker_lost_callback(settings, worker_id, reason="closed connection"):
+    dao = WorkerNodeDao(settings)
+    worker = dao.find_by_id(worker_id)
+
+    if worker is None:
+        logger.critical("dead worker {} not found".format(worker_id))
+        return
+
+    _handle_lost_worker_node(settings, worker, reason=reason)
+
+
+# assign available jobs to whoever is currently not working
+# triggered upon the following events
+# 1. client submitting a new job
+# 2. worker finishing a job
+def worker_schedule_job(settings):
+    conn_map = settings["WS_CONN_MAP"]
+    job_queue = settings["QUEUE"]
+
+    grading_job_dao = GradingJobDao(settings)
+    worker_node_dao = WorkerNodeDao(settings)
+
+    idle_workers = worker_node_dao.find_by_idleness()
 
-def _handle_lost_worker_node(settings, worker):
+    for idle_worker in idle_workers:
+        if idle_worker.use_ws and idle_worker.id in conn_map:
+            conn = conn_map[idle_worker.id]
```

Comment:
Nit: handle `conn == None` since it is being handled everywhere else.

Reply:
umm actually `conn_map[idle_worker.id]` is guaranteed to be a non-None value. Closed ones are deleted from the dict.